doc/DESIGN-AND-ROADMAP.md

```diff
@@ -61,6 +61,61 @@ Privacy Badger:
 
 #### Further Details
 
+Datastructures:
+
+action_map = { 'google.com': blocked, 'fonts.google.com', 'cookieblocked', 'apis.fonts.google.com': 'user_cookieblock', 'foo.tracker.net': 'allow', 'tracker.net': 'DNT', }
+snitch_map = {google.com: array('cooperq.com', 'noah.com', 'eff.org'), tracker.net: array(a.com, b.com, c.com)} 
+dnt_domains = array('tracker.net', 'dnt.eff.org')
+settings = {social_widgets = true, ...}
+cookie_block_list = "{'fonts.google.com': true, 'maps.google.com', true}"
+
+
+On Request():
+  check if privacy badger is enabled for this fqdn, if not then return
+  check if the fqdn is third party, if not then return
+  check action for the fqdn (described below)
+  if action is block cancel request 
+  if action is cookie_block strip headers 
+  if fqdn is nontracking (i.e check_action returned nothing) then do nothing
+  if action was noaction or user_* then async_check_tracking
+  if action is allow && count == 2 then blocking_check_tracking 
+    if check_tracking changed action then call check_action again
+    else do_nothing
+  asynch check_dnt(fqdn)
+  return
+
+Check_action(fqdn)  returns action
+  related_domains = array()
+  best_action == 'noaction'
+
+  for $domain in range(fqdn ... etld+1)
+    if action_map contains $domain
+      related_domains.push($domain)
+      
+
+    for each domain in related domains
+      if(score(domain.action) > score(best_action)
```

Comment:
push in this case is actually shift (add to the front of the array) and is so in the code, which makes it so that the array is ordered least to most specific which means that this works as expected (a user rule for gifs.twitter.com overrules twitter.com) but I will update the docs to reflect that.